### ISSUE TITLE
Option for blocking SSR or option for non-suspending refetch() fetchMore() of useSuspenseQuery()

### ISSUE THREAD

**Issue by peterbasar** [apollographql/apollo-client-nextjs#99]
Hello everyone,

I have question regarding SSR with NextJS and an app router.

I am working on big migration project moving from pages to app router.
Previously, SSR was handled by blocking initial page from sending until all queries have been finished (`qraphql()` hoc hooks that is).
This behavior was mostly handled like this: https://github.com/shshaw/next-apollo-ssr/blob/main/pages/_document.js
, with `getDataFromTree`.
(original issue for pages https://github.com/vercel/next.js/discussions/15736)

Now, I am on app router and I don't have a `_document.jsx` available. However I would still like to reproduce blocking SSR. I want to be able to migrate slowly since the amount of components I would have to refactor is large.

I was able to reproduce blocking SSR with https://github.com/apollographql/apollo-client-nextjs/ package and useSuspenseQuery hook by not wrapping it with `Suspense` component. However I encountered freezing issues.
Majority of these components are not designed with `Suspense` usage in mind and I basically have `useSuspenseQuery` next to other `useState` hooks. So I am unable to do something like this.
```
setVariables(...)
fetchMore(...)
```
I have other UI components depending on data that is set by `setVariables()` on user action, however when such action (option from dropdown is clicked) happens, the UI will not update and will suspend - freeze. 

I could redesign the whole 800 line component and other 10s or 100s of components/files (with `useSuspenseQuery`, `useBackgroundQuery` & `useReadQuery`) but the migration doesn't really make sense in this scenario. I want to add `Suspense` support gradually.

I want to know if it's possible to have this behavior (or very similar one) - any of these:

A)
1. Server waits (blocking) for all graphql useQuery hooks inside a tree (with option `{ssr: true}`) to get the data.
Queries marked with `{ssr: false}` are only executed on the browser. Potentially options like `{ssr: true, blocking: false}` or `{ssr:true, streaming: true}` etc.

2. Server responds only when all the data has been loaded.

B)
- Server doesn't wait for `useQuery` hooks (NextJS app router way I assume). But queries marked with `{ssr: true} `will be streamed.
Basically having `useQuery` hook + `{ssr: true}` have the same behavior as `useSuspenseQuery` but without actually suspending.

C)
- `useSuspenseQuery:refetch()` `useSuspenseQuery:fetchMore()` will have additional optional argument `{suspend: false}`. This will allow other parts of UI be updated and not freeze.

I hope I explained myself well enough. If you have any questions please reach out. I am quite new to SSR, Apollo and Suspense so my though process might not be very good.

**TLDR**
I want to reproduce blocking SSR on app router with NextJS but have the option to migrate to gradually to `Suspense` usage. Blocking SSR will only on the pages that contain blocking queries.
OR
I want to have option on `useSuspenseQuery:refetch()` or `useSuspenseQuery:fetchMore()` to not suspend.

**Comment by phryneas** [comment]
I think for the "blocking SSR", you're pretty much out of luck - you are migrating over to a Framework that has streaming SSR at it's core. 
Supporting that from our side would go against everything the App router stands for, and tbh., if you want to rely on that, you're a lot better off staying with the Pages router. 

If you migrate to the App router, I highly recommend that you also buy in on the new concepts it brings - and if that means a lot of code change just for migration's sake: 
You can also use the App and Pages router side-by-side, so nothing would force you to migrate over - you can leave your old code as it is - and write new code in the App router, with new concepts.

As for your "OR":
You can control how `useSuspenseQuery` variable changes, `refetch` or `fetchMore` suspend with React's [`useTransition` hook](https://react.dev/reference/react/useTransition). 
Suspending is React's job, so you need to use React primitives to take control over that. The only suspend that is set in stone is the suspend from the initial render.

**Comment by peterbasar** [comment]
Thank you very much @phryneas  for your insight.

I understand that the support for the non-streaming SSR is not necessarily something that would be desired in app router.
My goals with migration was to reproduce behavior of previous version 1:1 first before developing
improvements since the more time that is spent on migration branch, the more time the main branch is getting
further from its original version.

I think I can make use of Suspense component at least on initial render since the server response will
soon follow.
I was able to reproduce 'blocking' SSR by not wrapping `useSuspenseQuery()` with Suspense, but I assume that would just introduce unnecessary amount of waterfall query fetching on the server, and in turn, the whole migration would slow down the application.

Thank you for the hint with `useTransition()`. It really helped, and I am able to use the legacy-design version of the component without freezing the UI by wrapping `refetch()` with `startTransition()`. I will have to learn about it more.
(i was actually working on creating custom hook for switching between `useSuspenseQuery()` and `useQuery()` post SSR, so I can `refetch()` `fetch()` without component being suspended - which worked but this is better -> better but 1 thing I have still issue with - _next paragraph_)

One last thing, when I call `refetch()` that is wrapped by `startTransition()`, the data that is returned by the `useSuspenseQuery()` hook is old during the network fetching (even with 'no-cache' being set).
Because I noticed that the default behavior of `useQuery()` is that when the cache is set to 'no-cache' and we call `refetch()`, the `useQuery()` hook returns no data for the time of fetching.
Is that a desired behavior for `useSuspenseQuery()`?

I am asking because some older components rely on dataLength f.e. https://github.com/ankeetmaini/react-infinite-scroll-component.
I know that this is not necessarily a good way to handle loading state on `refetch()` - but that seemed to be a good source of information about the re-fetching status - data reset + data/no-data logic, and it is how `useQuery()` works. At the moment, the component is not suspended, and I have no idea about network status and cannot empty data provided by `useSuspendedQuery()`.
The only alternative solution I can think of is to do deep comparisons between old and new data as a way to get `refetch` status (however, this might cause issues when same data is returned even after different variables are used).

**Comment by phryneas** [comment]
> Because I noticed that the default behavior of useQuery() is that when the cache is set to 'no-cache' and we call refetch(), the useQuery() hook returns no data for the time of fetching.
Is that a desired behavior for useSuspenseQuery()?


Yes, the point of `useSuspenseQuery` (or suspense as a general concept) is that the component itself never renders in a "loading without data" state, so you always have data at your disposal.

What I could think of in your case would be wrapping `refetch` with a custom wrapper that also executes some `onRefetch` code to reset your additional information there.
